flowmachine/flowmachine/features/subscriber/contact_reciprocal.py

```diff
@@ -0,0 +1,388 @@
+    direction : {'in', 'out', 'both'}, default 'out'
+        Whether to consider calls made, received, or both. Defaults to 'out'.
+    exclude_self_calls : bool, default True
+        Set to false to *include* calls a subscriber made to themself
+    tables : str or list of strings, default 'all'
+        Can be a string of a single table (with the schema)
+        or a list of these. The keyword all is to select all
+        subscriber tables
+
+    Example
+    -------
+
+    >> s = ProportionEventReciprocal('2016-01-01', '2016-01-08')
+    >> s.get_dataframe()
+
+          subscriber       value
+    9vXy462Ej8V1kpWl         0.0
+    Q4mwVxpBOo7X2lb9         0.0
+    5jLW0EWeoyg6NQo3         0.0
+    QEoRM9vlkV18N4ZY         0.0
+    a76Ajyb9dmEYNd8L         0.0
+                 ...         ...
+    """
+
+    def __init__(
+        self,
+        start,
+        stop,
+        contact_reciprocal=None,
+        *,
+        direction="both",
+        subscriber_identifier="msisdn",
+        hours="all",
+        subscriber_subset=None,
+        tables="all",
+        exclude_self_calls=True,
+    ):
+
+        self.start = start
+        self.stop = stop
+        self.subscriber_identifier = subscriber_identifier
+        self.hours = hours
+        self.exclude_self_calls = exclude_self_calls
+        self.direction = direction
+        self.tables = tables
+
+        if self.direction == "both":
+            column_list = [self.subscriber_identifier, "msisdn", "msisdn_counterpart"]
+        else:
+            column_list = [
+                self.subscriber_identifier,
+                "msisdn",
+                "msisdn_counterpart",
+                "outgoing",
+            ]
+
+        self.unioned_query = EventsTablesUnion(
+            self.start,
+            self.stop,
+            tables=self.tables,
+            columns=column_list,
+            hours=hours,
+            subscriber_identifier=subscriber_identifier,
+            subscriber_subset=subscriber_subset,
+        )
+
+        if contact_reciprocal:
+            self.contact_reciprocal_query = contact_reciprocal
+        else:
+            contact_start = start
+            contact_stop = stop
+            contact_hours = hours
+            contact_tables = self.tables
+            self.contact_reciprocal_query = ContactReciprocal(
+                contact_start,
+                contact_stop,
+                tables=contact_tables,
+                hours=contact_hours,
+                exclude_self_calls=self.exclude_self_calls,
+                subscriber_subset=subscriber_subset,
+            )
+
+        super().__init__()
+
+    @property
+    def column_names(self):
+        return ["subscriber", "value"]
+
+    def _make_query(self):
+
+        filters = []
+        if self.direction != "both":
+            filters.append(
+                f"outgoing IS {'TRUE' if self.direction == 'out' else 'FALSE'}"
+            )
+        if self.exclude_self_calls:
```

Comment:
This is the non-testable bit, because the only reciprocal caller is the self-caller, right? Can we have a test for it _anyway_, showing that person is excluded and everybody else has a 0?